### ISSUE TITLE
Compression: Add threshold for compression 

### ISSUE THREAD

**Issue by spboyer** [aspnet/BasicMiddleware#185]
There are instances where compression is a detriment by comparison to the benefit. Where the time to compress and gzip payload is higher than the original payload.

ngnix defaults to all content with the `gzip_min_length` setting used to set the minimum length of a response to that will be gzipped. 

expressjs compression module is 1k by default, uses `threshold` option to set globally.

**Comment by Tratcher** [comment]
As content-length is optional and only set by a few components, this would require buffering the response body up to that threshold before enabling compression. That's doable, but it can cause un-expected delays for streaming scenarios.

**Comment by spboyer** [comment]
For the streaming scenarios the threshold should be set to 0, by passing the need to for check/routine.

**Comment by Tratcher** [comment]
Except that streaming done for some requests, where this setting would be for all requests.

**Comment by spboyer** [comment]
So is the use case for the compression middleware only for streaming? Considering that Kestrel is not a web facing server and should be proxied. So typically, compression would be handled by IIS, nginx or apache for requests.  Trying to find to use case, or is this in an effort to get to a public facing server?

**Comment by Tratcher** [comment]
Yes, the main use case is for when Kestrel or WebListener are used independently. Kestrel will eventually be safe for edge deployments, and WebListener already is.

Streams are not the main use case, but the middleware should not break streaming with too much buffering like the IIS implementation does.

**Comment by RehanSaeed** [comment]
When the static file middleware is being used, the size of the content is known. Can't that scenario be optimized?
